google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/ProtoSchemaConverter.java

```diff
@@ -29,51 +30,95 @@
 // protobuf::DescriptorProto
 // that can be reconstructed by the backend.
 public class ProtoSchemaConverter {
-  private static class StructName {
-    public String getName() {
-      count++;
-      return count == 1 ? "__ROOT__" : "__S" + count;
-    }
-
-    private int count = 0;
-  }
-
   private static ProtoSchema convertInternal(
-      Descriptor input, List<String> visitedTypes, StructName structName) {
+      Descriptor input,
+      List<String> visitedTypes,
+      List<String> enumTypes,
+      List<String> structTypes,
+      DescriptorProto.Builder rootProtoSchema) {
     DescriptorProto.Builder resultProto = DescriptorProto.newBuilder();
-    resultProto.setName(structName.getName());
+    if (rootProtoSchema == null) {
+      rootProtoSchema = resultProto;
+    }
+    String protoName = input.getFullName();
+    protoName = protoName.replace('.', '_');
+    resultProto.setName(protoName);
+    List<String> localEnumTypes = new ArrayList<String>();
     visitedTypes.add(input.getFullName());
     for (int i = 0; i < input.getFields().size(); i++) {
       FieldDescriptor inputField = input.getFields().get(i);
       FieldDescriptorProto.Builder resultField = inputField.toProto().toBuilder();
       if (inputField.getType() == FieldDescriptor.Type.GROUP
           || inputField.getType() == FieldDescriptor.Type.MESSAGE) {
-        if (visitedTypes.contains(inputField.getMessageType().getFullName())) {
-          throw new InvalidArgumentException(
-              "Recursive type is not supported:" + inputField.getMessageType().getFullName(),
-              null,
-              GrpcStatusCode.of(Status.Code.INVALID_ARGUMENT),
-              false);
+        String msgFullName = input.getFields().get(i).getMessageType().getFullName();
```

Comment:
inputField.getMessageType().getFullName() instead of input.getFields().get(i).getMessageType().getFullName();